src/metabase/models/permissions.clj

```diff
@@ -776,7 +829,65 @@
 
 ;;; ----------------------------------------------- Graph Updating Fns -----------------------------------------------
 
-(s/defn ^:private update-table-read-perms!
+(defn- download-permissions-set
+  [group-id]
+  (map :object
+       (db/select [Permissions :object]
+                  {:where [:and
+                           [:= :group_id group-id]
+                           [:or
+                            [:= :object (hx/literal "/")]
+                            [:like :object (hx/literal "/download/%")]]]})))
+
+(defn- download-permissions-level
+  [permissions-set db-id & [schema-name table-id]]
+  (cond
+   (set-has-full-permissions? permissions-set (feature-perms-path :download :full db-id schema-name table-id))
+   :full
+
+   (set-has-full-permissions? permissions-set (feature-perms-path :download :limited db-id schema-name table-id))
+   :limited
+
+   :else
+   :none))
+
+(s/defn update-native-download-permissions!
+  "To update native download permissions, we must read the list of tables in the database, and check the group's
```

Comment:
I'd add a comment explaining: the native download permission will be applied to the entire database, not a single table.